

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Module blaster.py &mdash; Leishmania_Project 0.1 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Module aesthetics.py" href="aesthetics.html" />
    <link rel="prev" title="Support" href="support.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> Leishmania_Project
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="support.html">Support</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Modules:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Module blaster.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="aesthetics.html">Module aesthetics.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="identifiers.html">Module identifiers.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="seq_modifier.html">Module seq_modifier.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="files_manager.html">Module files_manager.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="filters.html">Module filters.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="duplicates.html">Module duplicates.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="overlap.html">Module overlap.py</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Leishmania_Project</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Module blaster.py</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/blaster.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-modules.blaster">
<span id="module-blaster-py"></span><h1>Module blaster.py<a class="headerlink" href="#module-modules.blaster" title="Permalink to this headline">¶</a></h1>
<dl class="py function">
<dt class="sig sig-object py" id="modules.blaster.blastn_blaster">
<span class="sig-prename descclassname"><span class="pre">modules.blaster.</span></span><span class="sig-name descname"><span class="pre">blastn_blaster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">query_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dict_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">perc_identity</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#modules.blaster.blastn_blaster" title="Permalink to this definition">¶</a></dt>
<dd><p>This module calls for <cite>blastn</cite> in the BLAST<sup>R</sup> command line.
See <a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK279690">Command Line Application User Manual</a> for more information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>query_path</strong> (<em>string</em>) – Path to our FASTA query used in BLASTn against our database.</p></li>
<li><p><strong>dict_path</strong> (<em>string</em>) – Path to the FASTA file <code class="docutils literal notranslate"><span class="pre">query_path</span></code> will be launched to. In the same directory should be the BLAST data base created with <a class="reference internal" href="#modules.blaster.blastn_dic" title="modules.blaster.blastn_dic"><code class="xref py py-func docutils literal notranslate"><span class="pre">blastn_dic()</span></code></a>.</p></li>
<li><p><strong>outfile_path</strong> (<em>string</em>) – Path where the results will be saved.</p></li>
<li><p><strong>perc_identity</strong> (<em>string</em>) – Percent of sequence identity which we want to make the BLASTn. <strong>Important</strong>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A CSV file with the results of the BLASTn.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>CSV file</p>
</dd>
</dl>
<table class="docutils align-default" id="id1">
<caption><span class="caption-text">Arguments used in <code class="docutils literal notranslate"><span class="pre">blastn</span></code></span><a class="headerlink" href="#id1" title="Permalink to this table">¶</a></caption>
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Type</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">-word_size</span></code></p></td>
<td><p>integer</p></td>
<td><div class="line-block">
<div class="line">28 by default.</div>
<div class="line">BLAST search starts with finding a perfect sequence match of length given by this parameter.</div>
<div class="line">See <a class="reference external" href="https://www.metagenomics.wiki/tools/blast/default-word-size">world_size info</a> for more information.</div>
</div>
</td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">-query</span></code></p></td>
<td><p>string</p></td>
<td><p>Query file name.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">-db</span></code></p></td>
<td><p>string</p></td>
<td><p>BLAST database name.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">-out</span></code></p></td>
<td><p>string</p></td>
<td><p>Output file name.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">-perc_identity</span></code></p></td>
<td><p>integer</p></td>
<td><p>Percent identity cutoff.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">-outfmt</span></code></p></td>
<td><p>string</p></td>
<td><div class="line-block">
<div class="line">Alignment view options. Written inside in a nested string.</div>
<div class="line">Explained in <a class="reference internal" href="#outfmt-table"><span class="std std-ref">Table</span></a></div>
<div class="line">See <a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK279690">Command Line Application User Manual</a> and <a class="reference external" href="https://www.metagenomics.wiki/tools/blast/blastn-output-format-6s">Metagenomics BLASTn manual</a> for help.</div>
</div>
</td>
</tr>
</tbody>
</table>
<span id="outfmt-table"></span><table class="docutils align-default" id="id2">
<caption><span class="caption-text">Arguments used in <code class="docutils literal notranslate"><span class="pre">--outfmt</span></code> inside <code class="docutils literal notranslate"><span class="pre">blastn</span></code></span><a class="headerlink" href="#id2" title="Permalink to this table">¶</a></caption>
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">10</span></code></p></td>
<td><p>Comma-separated values.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">qseqid</span></code></p></td>
<td><div class="line-block">
<div class="line">Query Seq-id.</div>
<div class="line">ID of our query sequence.</div>
</div>
</td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">sseqid</span></code></p></td>
<td><div class="line-block">
<div class="line">Subject Seq-id.</div>
<div class="line">ID of our subject sequence.</div>
</div>
</td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">pident</span></code></p></td>
<td><p>Percentage of identical matches.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">length</span></code></p></td>
<td><p>Alignment length.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">qlen</span></code></p></td>
<td><p>Query sequence length.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">slen</span></code></p></td>
<td><p>Subject sequence length.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">mismatch</span></code></p></td>
<td><p>Number of mismatches.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">gapopen</span></code></p></td>
<td><p>Number of gap openings.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">qstart</span></code></p></td>
<td><p>Start of alignment in query.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">qend</span></code></p></td>
<td><p>End of alignment in query.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">sstart</span></code></p></td>
<td><p>Start of alignment in subject.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">send</span></code></p></td>
<td><p>End of alignment in subject.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">evalue</span></code></p></td>
<td><p>Expect value.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">bitscore</span></code></p></td>
<td><p>Bit score.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">sstrand</span></code></p></td>
<td><p>Subject strand.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">sseq</span></code></p></td>
<td><p>Aligned part of subject sequence.</p></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="modules.blaster.blastn_dic">
<span class="sig-prename descclassname"><span class="pre">modules.blaster.</span></span><span class="sig-name descname"><span class="pre">blastn_dic</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path_input</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#modules.blaster.blastn_dic" title="Permalink to this definition">¶</a></dt>
<dd><p>Creation af a BLAST database of our whole genome. It uses the BLAST<sup>R</sup> command line, see BLAST
<a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK279690">Command Line Application User Manual</a> for more information.</p>
<p>The generation of the properly database will be placed in the directory where <code class="docutils literal notranslate"><span class="pre">path_input</span></code> is.
It is recommended to use a dedicated folder to this FASTA file so the database is written next to it.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>path_input</strong> (<em>string</em>) – path to a FASTA file.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a BLAST database.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Muitiples files (<strong>.nhr</strong>, <strong>.nin</strong>, <strong>.nog</strong>, <strong>.nsd</strong>, <strong>.nsi</strong> and <strong>.nsq</strong> extensions)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="modules.blaster.repetitive_blaster">
<span class="sig-prename descclassname"><span class="pre">modules.blaster.</span></span><span class="sig-name descname"><span class="pre">repetitive_blaster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome_fasta</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">folder_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">naming_short</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_diff</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">numbering</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maximun_runs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#modules.blaster.repetitive_blaster" title="Permalink to this definition">¶</a></dt>
<dd><p>This function will iterate till a number of <code class="docutils literal notranslate"><span class="pre">maximun_runs</span></code> defined.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genome_fasta</strong> (<em>string</em>) – Path to our whole genome sequence in .fasta.</p></li>
<li><p><strong>path_input</strong> (<em>string</em>) – Path to the main CSV file where data will be filtered. Initially is a CSV file wich was output from <a class="reference internal" href="#modules.blaster.blastn_blaster" title="modules.blaster.blastn_blaster"><code class="xref py py-func docutils literal notranslate"><span class="pre">blastn_blaster()</span></code></a> alone to obtain the initial data.</p></li>
<li><p><strong>folder_path</strong> (<em>string</em>) – Path to a folder where the results will be placed. Subfolder will be created with the cromosome names.</p></li>
<li><p><strong>naming_short</strong> (<em>string</em>) – Label needed to read the ID of each cromosome in the .csv file. In the case of <strong>L. infantum</strong> for example, would be <em>LinJ</em> since the .csv file IDs are <em>LinJ.XX</em></p></li>
<li><p><strong>max_diff</strong> (<em>integer</em>) – Maximun proxomity value for the different sequences when they have to be grouped. <strong>Important</strong></p></li>
<li><p><strong>numbering</strong> (<em>integer</em>) – Indicates the number showed for the first program run. If it’s 0, then, the first run will be name 0.</p></li>
<li><p><strong>maximun_runs</strong> (<em>integer</em>) – Indicates the last iterative execution of function.</p></li>
</ul>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Pay special attention to the <code class="docutils literal notranslate"><span class="pre">max_diff</span></code> argument.</p>
</div>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="aesthetics.html" class="btn btn-neutral float-right" title="Module aesthetics.py" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="support.html" class="btn btn-neutral float-left" title="Support" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2023, Ronny Pacheco.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>