

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Welcome to Leishmania_Project’s documentation! &mdash; Leishmania_Project 0.1 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Installation" href="install.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="#" class="icon icon-home"> Leishmania_Project
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="support.html">Support</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">Leishmania_Project</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="#" class="icon icon-home"></a> &raquo;</li>
        
      <li>Welcome to Leishmania_Project’s documentation!</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/index.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="welcome-to-leishmania-project-s-documentation">
<h1>Welcome to Leishmania_Project’s documentation!<a class="headerlink" href="#welcome-to-leishmania-project-s-documentation" title="Permalink to this headline">¶</a></h1>
<div class="toctree-wrapper compound">
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="support.html">Support</a></li>
</ul>
</div>
</div>
<div class="section" id="program">
<h1>Program<a class="headerlink" href="#program" title="Permalink to this headline">¶</a></h1>
<span class="target" id="module-modules.files_manager"></span><dl class="py function">
<dt class="sig sig-object py" id="modules.files_manager.csv_creator">
<span class="sig-prename descclassname"><span class="pre">modules.files_manager.</span></span><span class="sig-name descname"><span class="pre">csv_creator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">writing_path_input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">writing_input</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#modules.files_manager.csv_creator" title="Permalink to this definition">¶</a></dt>
<dd><p>This function will create .csv files</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>writing_path_input</strong> (<em>string</em>) – Path where the .csv file will be created</p></li>
<li><p><strong>writing_input</strong> (<em>TO BE DONE</em>) – What info we want to add to the .csv file, normally a Matrix/Array 3D</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="modules.files_manager.csv_mixer">
<span class="sig-prename descclassname"><span class="pre">modules.files_manager.</span></span><span class="sig-name descname"><span class="pre">csv_mixer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path_input1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_input2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">writing_path_input</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#modules.files_manager.csv_mixer" title="Permalink to this definition">¶</a></dt>
<dd><p>Mixes two .csv files. Then, uses <a class="reference internal" href="#modules.files_manager.csv_creator" title="modules.files_manager.csv_creator"><code class="xref py py-func docutils literal notranslate"><span class="pre">csv_creator()</span></code></a> to generate a .csv with the mixed files.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path_input1</strong> (<em>string</em>) – Path to the first .csv to mix. It will take the first rows in the final mixed file.</p></li>
<li><p><strong>path_input2</strong> (<em>string</em>) – Path to the second .csv to mix. It will take rows adther the first .csv.</p></li>
<li><p><strong>writing_path_input</strong> (<em>string</em>) – path to both .csv files</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="modules.files_manager.fasta_creator">
<span class="sig-prename descclassname"><span class="pre">modules.files_manager.</span></span><span class="sig-name descname"><span class="pre">fasta_creator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path_input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fasta_output_path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#modules.files_manager.fasta_creator" title="Permalink to this definition">¶</a></dt>
<dd><p>HAY QUE MODIFICARLO, NECESITO QUE EL NUMERING SEA 01, 02, 03, 04…09, 10, 11. Y no como lo tengo ahora, que es 1, 2, 3, 4, 5…10, 11, 12.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="modules.files_manager.folder_creator">
<span class="sig-prename descclassname"><span class="pre">modules.files_manager.</span></span><span class="sig-name descname"><span class="pre">folder_creator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">options</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#modules.files_manager.folder_creator" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a Folder in the current path</p>
<dl class="field-list simple">
<dt class="field-odd">Paran options</dt>
<dd class="field-odd"><p>Name of the folder</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Folder in the current directory</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-modules.blaster"></span><dl class="py function">
<dt class="sig sig-object py" id="modules.blaster.blastn_blaster">
<span class="sig-prename descclassname"><span class="pre">modules.blaster.</span></span><span class="sig-name descname"><span class="pre">blastn_blaster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">query_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dict_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outfile_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">perc_identity</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#modules.blaster.blastn_blaster" title="Permalink to this definition">¶</a></dt>
<dd><p>This module calls for <cite>blastn</cite> in the BLAST<sup>R</sup> command line.
See <a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK279690">Command Line Application User Manual</a> for more information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>query_path</strong> (<em>string</em>) – Path to our .fasta query used in BLASTn against our database.</p></li>
<li><p><strong>dict_path</strong> (<em>string</em>) – Path to our database we made with <a class="reference internal" href="#modules.blaster.blastn_dic" title="modules.blaster.blastn_dic"><code class="xref py py-func docutils literal notranslate"><span class="pre">blastn_dic()</span></code></a>.</p></li>
<li><p><strong>outfile_path</strong> (<em>string</em>) – path where the results will be saved. <strong>Remember to write it with the .csv ending</strong>.</p></li>
<li><p><strong>perc_identity</strong> (<em>string</em>) – Percent oh identity which we want to make the BLASTn</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a .csv file with the results of the BLASTN</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>.csv file</p>
</dd>
</dl>
<table class="docutils align-default" id="id1">
<caption><span class="caption-text">Arguments used in <code class="docutils literal notranslate"><span class="pre">blastn</span></code></span><a class="headerlink" href="#id1" title="Permalink to this table">¶</a></caption>
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Type</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">-word_size</span></code></p></td>
<td><p>integer</p></td>
<td><div class="line-block">
<div class="line">28 by default.</div>
<div class="line">BLAST search starts with finding a perfect sequence match of length given by this parameter.</div>
<div class="line">See <a class="reference external" href="https://www.metagenomics.wiki/tools/blast/default-word-size">world_size info</a> for more information.</div>
</div>
</td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">-query</span></code></p></td>
<td><p>string</p></td>
<td><p>Query file name.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">-db</span></code></p></td>
<td><p>string</p></td>
<td><p>BLAST database name.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">-out</span></code></p></td>
<td><p>string</p></td>
<td><p>Output file name.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">-perc_identity</span></code></p></td>
<td><p>integer</p></td>
<td><p>Percent identity cutoff.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">-outfmt</span></code></p></td>
<td><p>string</p></td>
<td><div class="line-block">
<div class="line">Alignment view options. Written inside in a nested string.</div>
<div class="line">Explained in <a class="reference internal" href="#outfmt-table"><span class="std std-ref">Table</span></a></div>
<div class="line">See <a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK279690">Command Line Application User Manual</a> and <a class="reference external" href="https://www.metagenomics.wiki/tools/blast/blastn-output-format-6s">Metagenomics BLASTn manual</a> for help.</div>
</div>
</td>
</tr>
</tbody>
</table>
<span id="outfmt-table"></span><table class="docutils align-default" id="id2">
<caption><span class="caption-text">Arguments used in <code class="docutils literal notranslate"><span class="pre">--outfmt</span></code> inside <code class="docutils literal notranslate"><span class="pre">blastn</span></code></span><a class="headerlink" href="#id2" title="Permalink to this table">¶</a></caption>
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">10</span></code></p></td>
<td><p>Comma-separated values.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">qseqid</span></code></p></td>
<td><div class="line-block">
<div class="line">Query Seq-id.</div>
<div class="line">ID of our query sequence.</div>
</div>
</td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">sseqid</span></code></p></td>
<td><div class="line-block">
<div class="line">Subject Seq-id.</div>
<div class="line">ID of our subject sequence.</div>
</div>
</td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">pident</span></code></p></td>
<td><p>Percentage of identical matches.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">length</span></code></p></td>
<td><p>Alignment length.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">qlen</span></code></p></td>
<td><p>Query sequence length.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">slen</span></code></p></td>
<td><p>Subject sequence length.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">mismatch</span></code></p></td>
<td><p>Number of mismatches.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">gapopen</span></code></p></td>
<td><p>Number of gap openings.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">qstart</span></code></p></td>
<td><p>Start of alignment in query.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">qend</span></code></p></td>
<td><p>End of alignment in query.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">sstart</span></code></p></td>
<td><p>Start of alignment in subject.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">send</span></code></p></td>
<td><p>End of alignment in subject.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">evalue</span></code></p></td>
<td><p>Expect value.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">bitscore</span></code></p></td>
<td><p>Bit score.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">sstrand</span></code></p></td>
<td><p>Subject strand.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">sseq</span></code></p></td>
<td><p>Aligned part of subject sequence.</p></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="modules.blaster.blastn_dic">
<span class="sig-prename descclassname"><span class="pre">modules.blaster.</span></span><span class="sig-name descname"><span class="pre">blastn_dic</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path_input</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#modules.blaster.blastn_dic" title="Permalink to this definition">¶</a></dt>
<dd><p>Creation af a BLAST database of our whole genome. It uses the BLAST<sup>R</sup> command line, see BLAST
<a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK279690">Command Line Application User Manual</a> for more information.</p>
<p>The generation of the properly database will be placed in the <strong>.fasta directory</strong>.
It is recommended to use a dedicated folder to this genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>path_input</strong> (<em>string</em>) – path to where our file genome sequence .fasta is</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="modules.blaster.repetitive_blaster">
<span class="sig-prename descclassname"><span class="pre">modules.blaster.</span></span><span class="sig-name descname"><span class="pre">repetitive_blaster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome_fasta</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">folder_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">naming_short</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_diff</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">numbering</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maximun_runs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#modules.blaster.repetitive_blaster" title="Permalink to this definition">¶</a></dt>
<dd><p>This function will iterate till a number of <code class="docutils literal notranslate"><span class="pre">maximun_runs</span></code> defined.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genome_fasta</strong> (<em>string</em>) – Path to our genome sequence in .fasta.</p></li>
<li><p><strong>path_input</strong> (<em>string</em>) – Path to the .csv file where data will be filtered.</p></li>
<li><p><strong>folder_path</strong> (<em>string</em>) – Path to a folder where the results will be placed. Subfolder will be created with the cromosome names.</p></li>
<li><p><strong>naming_short</strong> (<em>string</em>) – Label needed to read the ID of each cromosome in the .csv file. In the case of <strong>L. infantum</strong> for example, would be <em>LinJ</em> since the .csv file IDs are <em>LinJ.XX</em></p></li>
<li><p><strong>max_diff</strong> (<em>integer</em>) – Maximun proxomity value for the different sequences when they have to be grouped. <strong>Important</strong></p></li>
<li><p><strong>numbering</strong> (<em>integer</em>) – Indicates the number showed for the first program run. If it’s 0, then, the first run will be name 0.</p></li>
<li><p><strong>maximun_runs</strong> (<em>integer</em>) – Indicates the last iterative execution of function.</p></li>
</ul>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Pay special attention to the <code class="docutils literal notranslate"><span class="pre">max_diff</span></code> argument.</p>
</div>
</dd></dl>

<span class="target" id="module-modules.identifiers"></span><dl class="py function">
<dt class="sig sig-object py" id="modules.identifiers.genome_specific_chromosome_main">
<span class="sig-prename descclassname"><span class="pre">modules.identifiers.</span></span><span class="sig-name descname"><span class="pre">genome_specific_chromosome_main</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path_input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome_ID</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">main_folder_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome_fasta</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">naming_short</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_diff</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#modules.identifiers.genome_specific_chromosome_main" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="modules.identifiers.specific_sequence_extractor">
<span class="sig-prename descclassname"><span class="pre">modules.identifiers.</span></span><span class="sig-name descname"><span class="pre">specific_sequence_extractor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path_input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome_ID</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">main_folder_path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#modules.identifiers.specific_sequence_extractor" title="Permalink to this definition">¶</a></dt>
<dd><p>First it reads a .csv file and takes the rows which include in the second column (the one with the chromosome IDs), the defined chromosome.
Then, it creates a .csv file  with all the selectes rows in a subfolder in a specified directory using <a class="reference internal" href="#modules.files_manager.folder_creator" title="modules.files_manager.folder_creator"><code class="xref py py-func docutils literal notranslate"><span class="pre">folder_creator()</span></code></a> and <a class="reference internal" href="#modules.files_manager.csv_creator" title="modules.files_manager.csv_creator"><code class="xref py py-func docutils literal notranslate"><span class="pre">csv_creator()</span></code></a>.</p>
<div class="admonition-example-of-use admonition">
<p class="admonition-title">Example of use</p>
<p>If <code class="docutils literal notranslate"><span class="pre">chromosome_ID</span> <span class="pre">=</span> <span class="pre">LinJ.02</span></code>, it will take all rows in the <strong>.csv</strong> where the second column contains <code class="docutils literal notranslate"><span class="pre">LinJ.02</span></code></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path_input</strong> (<em>string</em>) – Path where the .csv file we’ll use to read and filter data is.</p></li>
<li><p><strong>chromosome_ID</strong> (<em>string</em>) – Chromosome identifier, e.g., <em>LinJ.07</em>.</p></li>
<li><p><strong>main_folder_path</strong> (<em>string</em>) – Path where the results will be placed. It will create a subfolder with the chromosome_ID name.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A .csv file with the selected “chromosome_ID”.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>csv file</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-modules.seq_modifier"></span><dl class="py function">
<dt class="sig sig-object py" id="modules.seq_modifier.specific_sequence_1000nt">
<span class="sig-prename descclassname"><span class="pre">modules.seq_modifier.</span></span><span class="sig-name descname"><span class="pre">specific_sequence_1000nt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path_input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome_ID</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">main_folder_path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#modules.seq_modifier.specific_sequence_1000nt" title="Permalink to this definition">¶</a></dt>
<dd><p>This function will expand the selected sequence to 1000 nt and will create a CSV file with it.</p>
<p>It uses the function <code class="xref py py-func docutils literal notranslate"><span class="pre">csv_creator()</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path_input</strong> (<em>string</em>) – Path where the .csv file we’ll use to read and filter data is.</p></li>
<li><p><strong>chromosome_ID</strong> (<em>string</em>) – Chromosome identifier, e.g., <em>LinJ.07</em>.</p></li>
<li><p><strong>main_folder_path</strong> (<em>string</em>) – Path where the results will be placed. It will create a subfolder with the chromosome_ID name + “_1000nt.csv”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a csv file with the sufix “_1000nt.csv”</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>CSV file</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="modules.seq_modifier.specific_sequence_corrected">
<span class="sig-prename descclassname"><span class="pre">modules.seq_modifier.</span></span><span class="sig-name descname"><span class="pre">specific_sequence_corrected</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path_input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nucleotides1000_directory</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">main_folder_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome_ID</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#modules.seq_modifier.specific_sequence_corrected" title="Permalink to this definition">¶</a></dt>
<dd><div class="admonition danger">
<p class="admonition-title">Danger</p>
<p>NEED TO MODIFY IT BECAUSE OF SOME ERRORS</p>
</div>
<p>The main use of this function is to get the real “coordinates” of the sequence.</p>
<p>Remember we’ve got first, after the initial BLASNn, sequence expanded to 1000nt. The next part of the program was launching them one against each other to get a BLASTn which can provide us useful information about the real coordinates:</p>
<p>Now we’ve got the 1000 nt sequences with an aproximade location of where the SIDER2 are. And then, we’ve got the BLASTn (of them against each other) which can provide us with the real coordinates.</p>
<p>With this, this function will get the real coordinates.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path_input</strong> (<em>string</em>) – Path to the CSV file we’ll use to filter data. It’s the result from a BLASTn made between the expanded 1000nt sequences.</p></li>
<li><p><strong>nucleotides1000_directory</strong> (<em>string</em>) – <em>return</em> result from the function <a class="reference internal" href="#modules.seq_modifier.specific_sequence_1000nt" title="modules.seq_modifier.specific_sequence_1000nt"><code class="xref py py-func docutils literal notranslate"><span class="pre">specific_sequence_1000nt()</span></code></a></p></li>
<li><p><strong>main_folder_path</strong> (<em>string</em>) – Path where we’ll save the CSV data file to.</p></li>
<li><p><strong>chromosome_ID</strong> (<em>string</em>) – Identification of the cromosome, e.g., “LinJ.07”</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p></p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-modules.filters"></span><dl class="py function">
<dt class="sig sig-object py" id="modules.filters.filter_by_column">
<span class="sig-prename descclassname"><span class="pre">modules.filters.</span></span><span class="sig-name descname"><span class="pre">filter_by_column</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path_input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">column</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size_filter</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">writing_path_input</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#modules.filters.filter_by_column" title="Permalink to this definition">¶</a></dt>
<dd><p>This function will filter a CSV data depending on <code class="docutils literal notranslate"><span class="pre">length</span></code> (if we want to firlter by sequence length) or <code class="docutils literal notranslate"><span class="pre">percent</span></code> (if we want to filter by identity percent).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path_input</strong> (<em>string</em>) – Path to the .csv file we want to filter data.</p></li>
<li><p><strong>column</strong> (<em>string</em>) – Can be <code class="docutils literal notranslate"><span class="pre">length</span></code> (if we want to firlter by sequence length) or <code class="docutils literal notranslate"><span class="pre">percent</span></code> (if we want to filter by identity percent)</p></li>
<li><p><strong>size_filter</strong> (<em>integer</em>) – Number to filter dependint of the <strong>column</strong> argument.</p></li>
<li><p><strong>writing_path_input</strong> (<em>string</em>) – Path to the CSV file this function will create and save</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A CSV file with the dalta filtered depending on the <strong>column</strong> and <strong>size_filter</strong> argumetns.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>CSV file</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="install.html" class="btn btn-neutral float-right" title="Installation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2023, Ronny Pacheco.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>