

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Module overlap.py &mdash; Leishmania_Project 0.1 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Module duplicates.py" href="duplicates.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> Leishmania_Project
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="support.html">Support</a></li>
<li class="toctree-l1"><a class="reference internal" href="maps.html">Program Maps</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Modules:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="blaster.html">Module blaster.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="aesthetics.html">Module aesthetics.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="identifiers.html">Module identifiers.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="seq_modifier.html">Module seq_modifier.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="files_manager.html">Module files_manager.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="filters.html">Module filters.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="duplicates.html">Module duplicates.py</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Module overlap.py</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Leishmania_Project</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Module overlap.py</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/overlap.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-modules.overlap">
<span id="module-overlap-py"></span><h1>Module overlap.py<a class="headerlink" href="#module-modules.overlap" title="Permalink to this headline">¶</a></h1>
<dl class="py function">
<dt class="sig sig-object py" id="modules.overlap.genome_solap_by_pairs">
<span class="sig-prename descclassname"><span class="pre">modules.overlap.</span></span><span class="sig-name descname"><span class="pre">genome_solap_by_pairs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rows_to_filter</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome_fasta</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#modules.overlap.genome_solap_by_pairs" title="Permalink to this definition">¶</a></dt>
<dd><p>Partials overlaps will be filtered and joined.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>rows_to_filter</strong> (<em>array 3D</em>) – A Array 3D given by <a class="reference internal" href="#modules.overlap.genome_solap_main" title="modules.overlap.genome_solap_main"><code class="xref py py-func docutils literal notranslate"><span class="pre">genome_solap_main()</span></code></a> with all the supossed rows with overlaps.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="modules.overlap.genome_solap_location_filter">
<span class="sig-prename descclassname"><span class="pre">modules.overlap.</span></span><span class="sig-name descname"><span class="pre">genome_solap_location_filter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chromosome_rows</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#modules.overlap.genome_solap_location_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>For the DNA (“minus” or “plus” strand) we get the <em>start</em> and <em>end</em> coordinates.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>chromosome_rows</strong> (<em>python list</em>) – Given by <a class="reference internal" href="#modules.overlap.genome_solap_location_grouping" title="modules.overlap.genome_solap_location_grouping"><code class="xref py py-func docutils literal notranslate"><span class="pre">genome_solap_location_grouping()</span></code></a>. It’s a Python list with all the rows from a CSV of one specific chromosome.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>4 lists with the start and end coordinates for “minus” and “plus” DNA strands.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>python list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="modules.overlap.genome_solap_location_grouping">
<span class="sig-prename descclassname"><span class="pre">modules.overlap.</span></span><span class="sig-name descname"><span class="pre">genome_solap_location_grouping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chromosome_rows</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">DNA_sense</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_diff</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#modules.overlap.genome_solap_location_grouping" title="Permalink to this definition">¶</a></dt>
<dd><p>This function will analyze each position of coordinates (start and end) for each DNA strand. It will group them depending on nearness.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chromosome_rows</strong> (<em>python list</em>) – Given by <a class="reference internal" href="#modules.overlap.genome_solap_minmax" title="modules.overlap.genome_solap_minmax"><code class="xref py py-func docutils literal notranslate"><span class="pre">genome_solap_minmax()</span></code></a>. It’s a Python list with all the rows from a CSV of one specific chromosome.</p></li>
<li><p><strong>DNA_sense</strong> (<em>integer</em>) – Maximun proxomity value for the different sequences when they have to be grouped. <strong>Important</strong>.</p></li>
<li><p><strong>max_diff</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An Array 3D with the positions for “starts” and ends” in “groups” like <code class="docutils literal notranslate"><span class="pre">[[[group1</span> <span class="pre">starts][group2</span> <span class="pre">starts][...]]</span>&#160;&#160; <span class="pre">[[group1</span> <span class="pre">ends][group2</span> <span class="pre">ends][...]]]</span></code></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Array 3D</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="modules.overlap.genome_solap_main">
<span class="sig-prename descclassname"><span class="pre">modules.overlap.</span></span><span class="sig-name descname"><span class="pre">genome_solap_main</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome_fasta</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">naming_short</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_input</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_diff</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">writing_path_input</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#modules.overlap.genome_solap_main" title="Permalink to this definition">¶</a></dt>
<dd><p>Will call a variety of functions with the purpose of filtering overlaps in the data.</p>
<p>It uses <a class="reference internal" href="filters.html#modules.filters.chromosome_filter" title="modules.filters.chromosome_filter"><code class="xref py py-func docutils literal notranslate"><span class="pre">chromosome_filter()</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genome_fasta</strong> (<em>string</em>) – Path to our whole genome sequence in FASTA format.</p></li>
<li><p><strong>naming_short</strong> (<em>string</em>) – Label needed to read the ID of each cromosome in the .csv file. In the case of <strong>L. infantum</strong> for example, would be <em>LinJ</em> since the .csv file IDs are <em>LinJ.XX</em>.</p></li>
<li><p><strong>path_input</strong> (<em>string</em>) – Path to the CSV file we want to filter data. It’s the output file created by <a class="reference internal" href="blaster.html#modules.blaster.blastn_blaster" title="modules.blaster.blastn_blaster"><code class="xref py py-func docutils literal notranslate"><span class="pre">blastn_blaster()</span></code></a> and given here by <a class="reference internal" href="filters.html#modules.filters.global_filters_main" title="modules.filters.global_filters_main"><code class="xref py py-func docutils literal notranslate"><span class="pre">modules.filters.global_filters_main()</span></code></a>.</p></li>
<li><p><strong>max_diff</strong> (<em>intenger</em>) – Maximun proxomity value for the different sequences when they have to be grouped. <strong>Important</strong>.</p></li>
<li><p><strong>writing_path_input</strong> (<em>string</em>) – Path where the CSV file will be saved.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="modules.overlap.genome_solap_minmax">
<span class="sig-prename descclassname"><span class="pre">modules.overlap.</span></span><span class="sig-name descname"><span class="pre">genome_solap_minmax</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chromosome_rows</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_diff</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#modules.overlap.genome_solap_minmax" title="Permalink to this definition">¶</a></dt>
<dd><p>After doing the groupings with <a class="reference internal" href="#modules.overlap.genome_solap_location_grouping" title="modules.overlap.genome_solap_location_grouping"><code class="xref py py-func docutils literal notranslate"><span class="pre">genome_solap_location_grouping()</span></code></a> for each DNA strand. We simply get the minimun or maximun value depending if it’s a groupings of “starts” or “ends”, but mainly on the <em>strand</em> (because “plus” and “minus” strands have like inverted coordinates).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chromosome_rows</strong> (<em>python list</em>) – Given by <a class="reference internal" href="#modules.overlap.genome_solap_main" title="modules.overlap.genome_solap_main"><code class="xref py py-func docutils literal notranslate"><span class="pre">genome_solap_main()</span></code></a>. It’s a Python list with all the rows from a CSV of one specific chromosome.</p></li>
<li><p><strong>max_diff</strong> (<em>integer</em>) – Maximun proxomity value for the different sequences when they have to be grouped. <strong>Important</strong>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>4 list with the minimum or maximun values depending if it’s a groupings of “starts” or “ends”, but mainly on the <em>strand</em></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>4 python lists.</p>
</dd>
</dl>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="duplicates.html" class="btn btn-neutral float-left" title="Module duplicates.py" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2023, Ronny Pacheco.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>